File: Problems/No.290_Word_Pattern/word_pattern.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """

        # parameters
        strings = s.split()
        pattern_mapping = {}

        # start computing
        if (len(pattern) != len(strings)):
            return False
        else:
            for i in range(len(pattern)):
                if pattern[i] in pattern_mapping:
                    if pattern_mapping[pattern[i]] != strings[i]:
                        return False
                else:
                    if strings[i] in pattern_mapping.values():
                        return False
                    else:
                        pattern_mapping[pattern[i]] = strings[i]
        
        return True

File: Problems/No.290_Word_Pattern/word_pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        
        // parameters
        std::stringstream pString(s);
        std::unordered_map<char, string> pattern_matching;
        std::unordered_set<std::string> sCheck;
        std::vector<std::string> strings;
        std::string temp;

        // start computing
        while (pString >> temp) strings.push_back(temp);
        if (strings.size() != pattern.length()) return false;
        else {

            for (int i=0; i<pattern.size(); i++) {

                if (pattern_matching.find(pattern[i]) != pattern_matching.end()) { if (pattern_matching[pattern[i]] != strings[i]) return false; }
                else {

                    if (sCheck.count(strings[i]) > 0) return false;
                    else {

                        pattern_matching[pattern[i]] = strings[i];
                        sCheck.insert(strings[i]);

                    }

                }

            }

        }

        return true;

    }
};
